### PR TITLE
SCC 3467: Accurate item status filtering for ReCAP items

### PR DESCRIPTION
**Pull request by nonword**
This adds:
 - Adds `_fixItemStatusAggregation` step to ResponseMassager to incorporate SCSB item statuses in item status aggregation
 - All `findByUri` calls now start by optionally pre-retrieving the full set of item statuses from the SCSB API _if_ we need to build item aggregations or filter by item status. Note that we'll be making this call for _all_ bibs when these conditions are met even if we don't know if there are items in ReCAP. If the SCSB response is invalid or stalls (up to 3s) the response defaults to the indexed item statuses (i.e. itemAggregations and item_status filter may not accurately reflect scsb statuses)
 - Adds a new `EsClient` module to centralize ES queries (and simplify access)
 - Adds a new `ScsbClient` module to centralize SCSB queries (retiring underutilized and poorly name `ScsbClientWithRecap`)
 - Adds new `deepValue` util for safely checking deep properties of objects (a common task with ES queries and responses)
 - Add scsb fixture handling for new `getItemsAvailabilityForBnum` method, which involved refactoring scsb fixture handling to generalize the stubbing
 - changes the 'Not available' label to match what's in NYPL-Core (so that item status aggregations don't split unavailable items across two differently labeled buckets)
 - light linting and inline comments to files touched

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -342,9 +342,103 @@ exports.backslashes = (phrase, count = 1) => {
...
+  // Return next level:
+  return exports.deepValue(object[nextPath], paths, defaultTo)
+}
+
 exports.checkForNestedHitsAndSource = (resp) => {
-  return resp && resp.hits && resp.hits.hits && resp.hits.hits[0] &&
-    resp.hits.hits[0]._source && resp.hits.hits[0].inner_hits &&
-    resp.hits.hits[0].inner_hits.items && resp.hits.hits[0].inner_hits.items.hits &&
-    resp.hits.hits[0].inner_hits.items.hits.hits
+  return exports.deepValue(resp, 'hits.hits[0]._source') &&
```
Comment (charmingduchess) [inline lib/util.js:430]: 🤩

File: lib/resources.js [hunk 2]
```diff
@@ -524,6 +549,101 @@ module.exports = function (app, _private = null) {
...
+    const itemIsRecapClause = {
+      regexp: { 'items.holdingLocation.id': 'loc:rc.*' }
+    }
+    // Item's indexed status matches one of the filtered statuses:
+    const itemHasIndexedStatusClause = {
+      terms: { 'items.status.id': statuses }
+    }
+    // Item is marked Unavailable in SCSB:
+    const itemIsUnavailableInRecapClause = {
+      script: {
```
Comment (charmingduchess) [inline lib/resources.js:596]: does this need to be nested like this?

**Review by charmingduchess — COMMENTED** [review]
I think I've got a handle on this... in a nutshell, under certain circumstances that we need accurate status counts, you're doing two separate requests for non recap aggregations (to ES) and recap aggregations (to scsb, but formatted to mimic an es response).

File: lib/resources.js [hunk 3]
```diff
@@ -523,6 +549,104 @@ module.exports = function (app, _private = null) {
...
+   *  retrieved from SCSB.
+   *
+   *  @param {string[]} statuses - An array of statuses to filter on
+   *  @param {string[]} unavailableRecapBarcodes - An array of item barcodes
+   *    known to be unavailble
+   *
+   *  Returns an ES filter that matches the desired statuses, but also uses
+   *  the known unavailable items to override item statuses whose status may
+   *  not be correctly indexed (because ReCAP items that are unavailable will
+   *  be indexed as Available)
```
[the remark this answers is not in the document]
Comment (charmingduchess) [reply lib/resources.js]: looks good. i would still like to see inclusion of the fun fact that all partner items get indexed as available. to me it seems like an important distinction that they're not just out of date, but that they are systematically guaranteed to be so.

File: lib/resources.js (hunk 2 above)
Comment (nonword) [reply lib/resources.js:596]: Yes. I'm not sure why, but [that's how one adds script queries](https://www.elastic.co/guide/en/elasticsearch/reference/5.3/query-dsl-script-query.html)

**Review by charmingduchess — APPROVED** [review]
🤘

**Comment by nonword** [comment]
> I think I've got a handle on this... in a nutshell, under certain circumstances that we need accurate status counts, you're doing two separate requests for non recap aggregations (to ES) and recap aggregations (to scsb, but formatted to mimic an es response).

Yes, that sounds right. One SCSB call and either 1 or 2 ES calls. We're doing this:
 - fetch all scsb item statuses for the bib
 - do the primary ES query (optionally retrieving item aggregations)
   - if filtering on item_status, the query will override the status of recap items with the fetched scsb statuses
 - (in `availability_resolver`) if `itemAggregations` is present and bib has recap items, issue a second ES query to just get non-recap item statuses and merge that with the scsb-derived item statuses so that the `itemAggregations` correctly represents item statuses across recap and non-recap items

**Pull request merged**